aws/app/lambda/archive_form_responses/archiver.js

```diff
@@ -0,0 +1,147 @@
+
+    return {
+      statusCode: "SUCCESS"
+    };
+  }
+  catch (err) {
+    await reportErrorToSlack(new SNSClient({ region: REGION }), err.message);
+
+    return {
+      statusCode: "ERROR",
+      error: err.message,
+    };
+  }
+
+};
+
+async function archiveConsumedFormResponses(dynamoDb, s3Client) {
+  var lastReadResponse = null;
+
+  while (lastReadResponse !== undefined) {
+    let response = await getConsumedFormResponses(dynamoDb, lastReadResponse === null ? undefined : lastReadResponse);
+
+    if (response.formResponses.length > 0) {
+      for (const formResponse of response.formResponses) {
+        await saveFormResponseToS3(s3Client, formResponse.FormID.S, formResponse.SubmissionID.S, formResponse.FormSubmission.S);
+      }
+
+      await deleteFormResponsesFromDynamoDb(dynamoDb, response.formResponses);
+    }
+
+    lastReadResponse = response.lastReadResponse;
+  }
+}
+
+async function getConsumedFormResponses(dynamoDb, lastReadResponse = undefined) {
+  const queryCommandInput = {
+    TableName: DYNAMODB_VAULT_TABLE_NAME,
+    IndexName: DYNAMODB_VAULT_INDEX_NAME,
+    Limit: PROCESSING_BATCH_SIZE,
+    ExclusiveStartKey: lastReadResponse,
+    KeyConditionExpression: "Retrieved = :isRetrieved",
+    ExpressionAttributeValues: { ":isRetrieved": { N: "1" } },
+    ProjectionExpression: "FormID, SubmissionID, FormSubmission",
+  };
+
+  try {
+    const queryCommandOutput = await dynamoDb.send(new QueryCommand(queryCommandInput));
+
+    return {
+      formResponses: queryCommandOutput.Items,
+      lastReadResponse: queryCommandOutput.LastEvaluatedKey,
+    };
+  }
+  catch (err) {
+    throw new Error("Failed to retrieve consumed form responses.")
+  }
+}
+
+async function saveFormResponseToS3(s3Client, formID, submissionID, formResponse) {
+  const putObjectCommandInput = {
+    Bucket: ARCHIVING_S3_BUCKET,
+    Body: formResponse,
+    Key: `${new Date().toISOString().slice(0, 10)}/${formID}/${submissionID}.json`,
+  };
+
+  try {
+    await s3Client.send(new PutObjectCommand(putObjectCommandInput));
+  }
+  catch (err) {
+    throw new Error(`Failed to save form response to S3 (SubmissionID = ${submissionID})`);
+  }
+}
+
+async function deleteFormResponsesFromDynamoDb(dynamoDb, formResponses) {
+  const deleteRequests = formResponses.reduce((accumulator, currentValue) => {
+    const deleteRequest = {
+      "DeleteRequest": {
+        "Key": {
+          "FormID": {
+            "S": currentValue.FormID.S
+          },
+          "SubmissionID": {
+            "S": currentValue.SubmissionID.S
+          }
+        }
+      }
+    };
+    return [...accumulator, ...[deleteRequest]];
+  }, []);
+
+  const batchWriteItemCommandInput = {
+    RequestItems: {
+      [DYNAMODB_VAULT_TABLE_NAME]: deleteRequests
+    },
+  };
+
+  try {
+    await dynamoDb.send(new BatchWriteItemCommand(batchWriteItemCommandInput));
+  }
+  catch (err) {
+    throw new Error("Failed to delete form responses from DynamoDB.")
```

Comment:
Just wondering if it might be better to send the raw error data through to slack or in the very least log it so we have some idea why the error happened

Reply:
I agree. I am adding the `err.message` to my custom error message.